### PR TITLE
Propagate phase changes from executions/subinstallations to installations

### PR DESCRIPTION
**Pull request by Diaphteiros**
**How to categorize this PR?**

/area robustness
/kind enhancement
/priority 3

See also #276

**Which issue(s) this PR fixes**:
This is part of #239 / #275

**Release note**:

```other operator
Phase changes in executions and subinstallations are now properly propagated to their owning installation, even if the installation was already in a final phase and there weren't any changes to the spec.
```

### CODE REVIEW

File: pkg/landscaper/controllers/installations/controller.go [hunk 1]
```diff
@@ -131,6 +131,15 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	}
 
 	if lsv1alpha1helper.IsCompletedInstallationPhase(inst.Status.Phase) && inst.Status.ObservedGeneration == inst.Generation {
+		// check whether the current phase does still match the combined phase of subinstallations and executions
+		op, err := c.initPrerequisites(ctx, inst)
```
Comment (schrodit) [inline pkg/landscaper/controllers/installations/controller.go:135]: why do we need to init all prerequisites?
This is a expensive function that resolves the component descriptor and blueprint.
I think we should rather only init them when we need to go to succeeded phase?

File: pkg/landscaper/installations/helper.go [hunk 2]
```diff
@@ -322,3 +323,39 @@ func StringMapToRawMessageMap(m map[string]string) map[string]json.RawMessage {
...
+	}
+	phases := []lsv1alpha1.ComponentInstallationPhase{}
+	if exec != nil {
+		phases = append(phases, lsv1alpha1.ComponentInstallationPhase(exec.Status.Phase))
+	}
+	for _, sub := range subinsts {
+		phases = append(phases, sub.Status.Phase)
+	}
+	cp := lsv1alpha1helper.CombinedInstallationPhase(phases...)
+	if o.Inst.Info.Status.Phase != cp {
```
Comment (schrodit) [inline pkg/landscaper/installations/helper.go:351]: I think we also need to update the exports and notify dependent installations if the installation's status changes to succeeded

**Review by schrodit — CHANGES_REQUESTED** [review]

File: pkg/landscaper/controllers/installations/controller.go (hunk 1 above)
Comment (Diaphteiros) [reply pkg/landscaper/controllers/installations/controller.go]: I removed the requirement for the operation

File: pkg/landscaper/controllers/installations/controller.go [hunk 3]
```diff
@@ -131,6 +131,11 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	}
 
 	if lsv1alpha1helper.IsCompletedInstallationPhase(inst.Status.Phase) && inst.Status.ObservedGeneration == inst.Generation {
+		// check whether the current phase does still match the combined phase of subinstallations and executions
+		err := installations.HandleSubComponentPhaseChanges(ctx, c.Client(), inst, logger)
+		if err != nil {
```
Comment (schrodit) [inline pkg/landscaper/controllers/installations/controller.go:136]:
```suggestion
		if err := installations.HandleSubComponentPhaseChanges(ctx, c.Client(), inst, logger); err != nil {
```

File: pkg/landscaper/installations/helper.go [hunk 4]
```diff
@@ -322,3 +323,38 @@ func StringMapToRawMessageMap(m map[string]string) map[string]json.RawMessage {
...
+		return fmt.Errorf("error fetching subinstallations for installation %s/%s: %w", inst.Namespace, inst.Name, err)
+	}
+	phases := []lsv1alpha1.ComponentInstallationPhase{}
+	if exec != nil {
+		phases = append(phases, lsv1alpha1.ComponentInstallationPhase(exec.Status.Phase))
+	}
+	for _, sub := range subinsts {
+		phases = append(phases, sub.Status.Phase)
+	}
+	cp := lsv1alpha1helper.CombinedInstallationPhase(phases...)
```
Comment (schrodit) [inline pkg/landscaper/installations/helper.go:349]:
```suggestion
	var phases []lsv1alpha1.ComponentInstallationPhase
	if execRef != nil {
		exec := &lsv1alpha1.Execution{}
		err := kubeClient.Get(ctx, execRef.NamespacedName(), exec)
		if err != nil {
			return fmt.Errorf("error getting execution for installation %s/%s: %w", inst.Namespace, inst.Name, err)
		}
		phases = append(phases, lsv1alpha1.ComponentInstallationPhase(exec.Status.Phase))
	}
	subinsts, err := listSubinstallations(ctx, kubeClient, inst)
	if err != nil {
		return fmt.Errorf("error fetching subinstallations for installation %s/%s: %w", inst.Namespace, inst.Name, err)
	}
	for _, sub := range subinsts {
		phases = append(phases, sub.Status.Phase)
	}
	cp := lsv1alpha1helper.CombinedInstallationPhase(phases...)
```

**Review by schrodit — CHANGES_REQUESTED** [review]

File: pkg/landscaper/installations/helper.go [hunk 5]
```diff
@@ -449,3 +450,57 @@ func ResolveConfigMapReference(ctx context.Context, kubeClient client.Client, co
 
 	return completeData, data, cm.Generation, nil
 }
+
+// HandleSubComponentPhaseChanges updates the phase of the given installation, if its phase doesn't match the combined phase of its subinstallations/executions anymore
+func HandleSubComponentPhaseChanges(ctx context.Context, kubeClient client.Client, inst *lsv1alpha1.Installation, getOperation func(context.Context, *lsv1alpha1.Installation) (*Operation, error), getExports func(context.Context, *Operation) ([]*dataobjects.DataObject, []*dataobjects.Target, error), logger logr.Logger) error {
```
Comment (schrodit) [inline pkg/landscaper/installations/helper.go:455]: this looks unreadable on a non-widescreen.
Can you make it more readable like
```suggestion
func HandleSubComponentPhaseChanges(
  ctx context.Context,
  kubeClient client.Client,
  inst *lsv1alpha1.Installation,
  getOperation func(context.Context, *lsv1alpha1.Installation) (*Operation, error),
  getExports func(context.Context, *Operation) ([]*dataobjects.DataObject, []*dataobjects.Target, error), logger logr.Logger) error {
```

File: pkg/landscaper/installations/helper.go [hunk 6]
```diff
@@ -449,3 +450,57 @@ func ResolveConfigMapReference(ctx context.Context, kubeClient client.Client, co
...
+	if inst.Status.Phase != cp {
+		// Phase is completed but doesn't fit to the deploy items' phases
+
+		// get operation
+		instOp, err := getOperation(ctx, inst)
+		if err != nil {
+			return fmt.Errorf("unable to construct operation for installation %s/%s: %w", inst.Namespace, inst.Name, err)
+		}
+		logger.V(5).Info("execution phase mismatch", "phase", string(inst.Status.Phase), "combinedPhase", string(cp))
+		err = instOp.UpdateInstallationStatus(ctx, inst, cp)
```
Comment (schrodit) [inline pkg/landscaper/installations/helper.go:483]: shoudn't we first update the export and then update the status?

File: pkg/landscaper/installations/helper.go [hunk 7]
```diff
@@ -449,3 +450,57 @@ func ResolveConfigMapReference(ctx context.Context, kubeClient client.Client, co
...
+		if err != nil {
+			return fmt.Errorf("error updating installation status for installation %s/%s: %w", inst.Namespace, inst.Name, err)
+		}
+
+		// recompute exports
+		dataExports, targetExports, err := getExports(ctx, instOp)
```
Comment (schrodit) [inline pkg/landscaper/installations/helper.go:489]: this is only needed if we are in succeeded state.
With that code we would also update the exports during every other state e.g. `Progressing` or `Failed`

**Review by schrodit — CHANGES_REQUESTED** [review]

File: pkg/landscaper/installations/helper.go [hunk 8]
```diff
@@ -449,3 +450,70 @@ func ResolveConfigMapReference(ctx context.Context, kubeClient client.Client, co
...
 }
+
+// HandleSubComponentPhaseChanges updates the phase of the given installation, if its phase doesn't match the combined phase of its subinstallations/executions anymore
+func HandleSubComponentPhaseChanges(
+	ctx context.Context,
+	kubeClient client.Client,
+	inst *lsv1alpha1.Installation,
+	getOperation func(context.Context, *lsv1alpha1.Installation) (*Operation, error),
+	getExports func(context.Context, *Operation) ([]*dataobjects.DataObject, []*dataobjects.Target, error),
+	logger logr.Logger) error {
```
Comment (schrodit) [inline pkg/landscaper/installations/helper.go:461]: nit: in other functions we pass the logger as second argument after the context.
```suggestion
	ctx context.Context,
	log logr.Logger,
	kubeClient client.Client,
	inst *lsv1alpha1.Installation,
	getOperation func(context.Context, *lsv1alpha1.Installation) (*Operation, error),
	getExports func(context.Context, *Operation) ([]*dataobjects.DataObject, []*dataobjects.Target, error)) error {
```

**Review by schrodit — CHANGES_REQUESTED** [review]
just a minor nit otherwise lgmt.

Can you squash all your commits?

File: pkg/landscaper/installations/helper.go (hunk 8 above)
Comment (Diaphteiros) [reply pkg/landscaper/installations/helper.go]: Ah, I wasn't sure about that. I changed it. Will squash the commits during merge.

**Review by schrodit — APPROVED** [review]
/lgtm
/merge squash

**Pull request merged**